File: fortran/bench/linpack.c
/* The same LINPACK kernel as linpack.f, written to mirror it exactly:
 * identical loop structure, identical COLUMN-MAJOR indexing (a[(i-1) +
 * (j-1)*lda]), identical by-reference argument passing.  The point is
 * to compare CODE GENERATION for the same work, not two different
 * algorithms or two different memory access patterns. */

static void daxpy(int n, double da, double *dx, double *dy) {
    int i;
    if (n <= 0) return;
    if (da == 0.0) return;
    for (i = 1; i <= n; i++) dy[i-1] = dy[i-1] + da * dx[i-1];
}

static void dscal(int n, double da, double *dx) {
    int i;
    if (n <= 0) return;
    for (i = 1; i <= n; i++) dx[i-1] = da * dx[i-1];
}

static double dabs_(double v) { return v < 0.0 ? -v : v; }

static int idamax(int n, double *dx) {
    double dmax;
    int i, r;
    if (n < 1) return 0;
    if (n == 1) return 1;
    r = 1;
    dmax = dabs_(dx[0]);
    for (i = 2; i <= n; i++) {
        if (dabs_(dx[i-1]) <= dmax) continue;
        r = i;
        dmax = dabs_(dx[i-1]);
    }
    return r;
}

#define A(i,j) a[((i)-1) + ((j)-1)*lda]

static void dgefa(double *a, int lda, int n, int *ipvt, int *info) {
    double t;
    int j, k, kp1, l, nm1;
    *info = 0;
    nm1 = n - 1;
    if (nm1 >= 1) {
        for (k = 1; k <= nm1; k++) {
            kp1 = k + 1;
            l = idamax(n-k+1, &A(k,k)) + k - 1;
            ipvt[k-1] = l;
            if (A(l,k) == 0.0) { *info = k; continue; }
            if (l != k) { t = A(l,k); A(l,k) = A(k,k); A(k,k) = t; }
            t = -1.0 / A(k,k);
            dscal(n-k, t, &A(k+1,k));
            for (j = kp1; j <= n; j++) {
                t = A(l,j);
                if (l != k) { A(l,j) = A(k,j); A(k,j) = t; }
                daxpy(n-k, t, &A(k+1,k), &A(k+1,j));
            }
        }
    }
    ipvt[n-1] = n;
    if (A(n,n) == 0.0) *info = n;
}

static void dgesl(double *a, int lda, int n, int *ipvt, double *b) {
    double t;
    int k, kb, l, nm1;
    nm1 = n - 1;
    if (nm1 >= 1) {
        for (k = 1; k <= nm1; k++) {
            l = ipvt[k-1];
            t = b[l-1];
            if (l != k) { b[l-1] = b[k-1]; b[k-1] = t; }
            daxpy(n-k, t, &A(k+1,k), &b[k]);
        }
    }
    for (kb = 1; kb <= n; kb++) {
        k = n + 1 - kb;
        b[k-1] = b[k-1] / A(k,k);
        t = -b[k-1];
        daxpy(k-1, t, &A(1,k), &b[0]);
    }
}

static double a_[64*64];
static double aa_[64*64];
static double b_[64];
static int    ipvt_[64];

int main(void) {
    int lda = 64, n = 64, reps = 400;
    int i, j, r, info;
    double resid;
    double *a = aa_;

    for (j = 1; j <= n; j++) {
        for (i = 1; i <= n; i++) A(i,j) = 1.0 / (double)(i + j);
        A(j,j) = A(j,j) + (double)n;
    }
    for (r = 1; r <= reps; r++) {
        double *src = aa_;
        a = a_;
        for (j = 1; j <= n; j++)
            for (i = 1; i <= n; i++)
                a_[(i-1) + (j-1)*lda] = src[(i-1) + (j-1)*lda];
        for (i = 1; i <= n; i++) {
            b_[i-1] = 0.0;
            for (j = 1; j <= n; j++) b_[i-1] = b_[i-1] + src[(i-1) + (j-1)*lda];
        }
        dgefa(a_, lda, n, ipvt_, &info);
        if (info != 0) return 91;
        dgesl(a_, lda, n, ipvt_, b_);
    }
    resid = 0.0;
    for (i = 1; i <= n; i++) {
        double d = dabs_(b_[i-1] - 1.0);
        if (d > resid) resid = d;
    }
    if (resid > 1e-8) return 92;
    return 0;
}

File: fortran/bench/linpack.f
C     LINPACK benchmark kernel, sized to dominate startup cost.
C     Factors and solves an N x N system REPS times.  Self-checking:
C     exits 0 only if every solution component is 1 to within 1e-8.
      PROGRAM LINBEN
      INTEGER LDA, N, REPS, IPVT(64), INFO, I, J, R
      DOUBLE PRECISION A(64,64), AA(64,64), B(64), RESID
      LDA = 64
      N = 64
      REPS = 400
      DO 20 J = 1, N
         DO 10 I = 1, N
            AA(I,J) = 1.0D0 / DBLE(I + J)
   10    CONTINUE
         AA(J,J) = AA(J,J) + DBLE(N)
   20 CONTINUE
      DO 70 R = 1, REPS
         DO 40 J = 1, N
            DO 30 I = 1, N
               A(I,J) = AA(I,J)
   30       CONTINUE
   40    CONTINUE
         DO 60 I = 1, N
            B(I) = 0.0D0
            DO 50 J = 1, N
               B(I) = B(I) + AA(I,J)
   50       CONTINUE
   60    CONTINUE
         CALL DGEFA(A, LDA, N, IPVT, INFO)
         IF (INFO .NE. 0) STOP 91
         CALL DGESL(A, LDA, N, IPVT, B)
   70 CONTINUE
      RESID = 0.0D0
      DO 80 I = 1, N
         RESID = DMAX1(RESID, DABS(B(I) - 1.0D0))
   80 CONTINUE
      IF (RESID .GT. 1.0D-8) STOP 92
      STOP 0
      END

      SUBROUTINE DAXPY(N, DA, DX, INCX, DY, INCY)
      DOUBLE PRECISION DX(1), DY(1), DA
      INTEGER I, INCX, INCY, N
      IF (N .LE. 0) RETURN
      IF (DA .EQ. 0.0D0) RETURN
      DO 10 I = 1, N
         DY(I) = DY(I) + DA*DX(I)
   10 CONTINUE
      RETURN
      END

      SUBROUTINE DSCAL(N, DA, DX, INCX)
      DOUBLE PRECISION DA, DX(1)
      INTEGER I, INCX, N
      IF (N .LE. 0) RETURN
      DO 10 I = 1, N
         DX(I) = DA*DX(I)
   10 CONTINUE
      RETURN
      END

      INTEGER FUNCTION IDAMAX(N, DX, INCX)
      DOUBLE PRECISION DX(1), DMAX
      INTEGER I, INCX, N
      IDAMAX = 0
      IF (N .LT. 1) RETURN
      IDAMAX = 1
      IF (N .EQ. 1) RETURN
      DMAX = DABS(DX(1))
      DO 10 I = 2, N
         IF (DABS(DX(I)) .LE. DMAX) GO TO 10
         IDAMAX = I
         DMAX = DABS(DX(I))
   10 CONTINUE
      RETURN
      END

      SUBROUTINE DGEFA(A, LDA, N, IPVT, INFO)
      INTEGER LDA, N, IPVT(1), INFO
      DOUBLE PRECISION A(LDA,1), T
      INTEGER IDAMAX, J, K, KP1, L, NM1
      INFO = 0
      NM1 = N - 1
      IF (NM1 .LT. 1) GO TO 70
      DO 60 K = 1, NM1
         KP1 = K + 1
         L = IDAMAX(N-K+1, A(K,K), 1) + K - 1
         IPVT(K) = L
         IF (A(L,K) .EQ. 0.0D0) GO TO 40
         IF (L .EQ. K) GO TO 10
         T = A(L,K)
         A(L,K) = A(K,K)
         A(K,K) = T
   10    CONTINUE
         T = -1.0D0/A(K,K)
         CALL DSCAL(N-K, T, A(K+1,K), 1)
         DO 30 J = KP1, N
            T = A(L,J)
            IF (L .EQ. K) GO TO 20
            A(L,J) = A(K,J)
            A(K,J) = T
   20       CONTINUE
            CALL DAXPY(N-K, T, A(K+1,K), 1, A(K+1,J), 1)
   30    CONTINUE
         GO TO 50
   40    CONTINUE
         INFO = K
   50    CONTINUE
   60 CONTINUE
   70 CONTINUE
      IPVT(N) = N
      IF (A(N,N) .EQ. 0.0D0) INFO = N
      RETURN
      END

      SUBROUTINE DGESL(A, LDA, N, IPVT, B)
      INTEGER LDA, N, IPVT(1)
      DOUBLE PRECISION A(LDA,1), B(1), T
      INTEGER K, KB, L, NM1
      NM1 = N - 1
      IF (NM1 .LT. 1) GO TO 30
      DO 20 K = 1, NM1
         L = IPVT(K)
         T = B(L)
         IF (L .EQ. K) GO TO 10
         B(L) = B(K)
         B(K) = T
   10    CONTINUE
         CALL DAXPY(N-K, T, A(K+1,K), 1, B(K+1), 1)
   20 CONTINUE
   30 CONTINUE
      DO 40 KB = 1, N
         K = N + 1 - KB
         B(K) = B(K)/A(K,K)
         T = -B(K)
         CALL DAXPY(K-1, T, A(1,K), 1, B(1), 1)
   40 CONTINUE
      RETURN
      END
